src/replica/replica_learn.cpp

```diff
@@ -1549,21 +1561,23 @@ error_code replica::apply_learned_state_from_private_log(learn_state &state)
     prepare_list plist(this,
                        _app->last_committed_decree(),
                        _options->max_mutation_count_in_prepare_list,
-                       [this, duplicating](mutation_ptr &mu) {
+                       [this, duplicating, step_back](mutation_ptr &mu) {
                            if (mu->data.header.decree == _app->last_committed_decree() + 1) {
                                // TODO: assign the returned error_code to err and check it
                                _app->apply_mutation(mu);
 
                                // appends logs-in-cache into plog to ensure them can be duplicated.
-                               if (duplicating) {
+                               // if current case is step back, it means the logs has been reserved
+                               // through `reset_form` above
+                               if (duplicating && !step_back) {
                                    _private_log->append(
                                        mu, LPC_WRITE_REPLICATION_LOG_COMMON, &_tracker, nullptr);
                                }
                            }
                        });
 
     err = mutation_log::replay(state.files,
-                               [&plist](int log_length, mutation_ptr &mu) {
+                               [step_back, duplicating, &plist](int log_length, mutation_ptr &mu) {
```

Comment:
It seems that `step_back` and `duplicating` are not used in callback function of replay, why do you add them here?

Reply:
deleted